upstream/upstream.go

```diff
@@ -922,28 +922,31 @@ func (u *Upstream) detectFeatures(ctx context.Context) error {
 		if cfg.Evm == nil {
 			cfg.Evm = &common.EvmUpstreamConfig{}
 		}
-		if cfg.Evm.ChainId == 0 {
-			nid, err := u.EvmGetChainId(ctx)
-			if err != nil {
-				return common.NewErrUpstreamClientInitialization(
-					&common.BaseError{
-						Code:  "ErrUpstreamChainIdDetectionFailed",
-						Cause: err,
-					},
-					u,
-				)
-			}
-			cfg.Evm.ChainId, err = strconv.ParseInt(nid, 10, 64)
-			if err != nil {
-				return common.NewErrUpstreamClientInitialization(
-					&common.BaseError{
-						Code:  "ErrUpstreamChainIdDetectionFailed",
-						Cause: err,
-					},
-					u,
-				)
+		// Always detect the actual chainId from the endpoint to avoid cross-network misconfiguration.
+		// If user provided a chainId and it mismatches, prefer the detected value to ensure correct routing/metrics.
```

Comment:
I think we should actually explicitly fail here, so user does a double check why their expected chain ID doesn't match... if they want auto-detection they can just leave it as is.